src/mc.rs

```diff
@@ -36,46 +35,50 @@ use crate::include::common::bitdepth::{bpc_fn, BPC};
 
 #[inline(never)]
 unsafe fn put_rust<BD: BitDepth>(
-    dst: *mut BD::Pixel,
-    dst_stride: usize,
-    src: *const BD::Pixel,
-    src_stride: usize,
+    mut dst: *mut BD::Pixel,
+    dst_stride: isize,
+    mut src: *const BD::Pixel,
+    src_stride: isize,
     w: usize,
     h: usize,
 ) {
-    let [dst_len, src_len] =
-        [dst_stride, src_stride].map(|stride| if h == 0 { 0 } else { stride * (h - 1) + w });
-    let dst = slice::from_raw_parts_mut(dst, dst_len);
-    let src = slice::from_raw_parts(src, src_len);
-    for (dst, src) in iter::zip(dst.chunks_mut(dst_stride), src.chunks(src_stride)) {
-        BD::pixel_copy(dst, src, w);
+    for _ in 0..h {
+        BD::pixel_copy(
+            slice::from_raw_parts_mut(dst, w),
+            slice::from_raw_parts(src, w),
+            w,
+        );
+
+        dst = dst.offset(dst_stride);
+        src = src.offset(src_stride);
     }
 }
 
 #[inline(never)]
 unsafe fn prep_rust<BD: BitDepth>(
     tmp: *mut i16,
-    src: *const BD::Pixel,
-    src_stride: usize,
+    mut src: *const BD::Pixel,
+    src_stride: isize,
     w: usize,
     h: usize,
     bd: BD,
 ) {
     let tmp = slice::from_raw_parts_mut(tmp, w * h);
-    let src = slice::from_raw_parts(src, if h == 0 { 0 } else { src_stride * (h - 1) + w });
     let intermediate_bits = bd.get_intermediate_bits();
-    for (tmp, src) in iter::zip(tmp.chunks_exact_mut(w), src.chunks(src_stride)).take(h) {
-        for (tmp, src) in iter::zip(tmp, &src[..w]) {
-            *tmp = (((*src).as_::<i32>() << intermediate_bits) - (BD::PREP_BIAS as i32)) as i16;
+    for tmp in tmp.chunks_exact_mut(w).take(h) {
+        for x in 0..w {
+            tmp[x] =
+                (((*src.add(x)).as_::<i32>() << intermediate_bits) - (BD::PREP_BIAS as i32)) as i16
```

Comment:
Can we still make `src` a slice for the inner loop?

Reply:
I don't see much point to that, we're just replacing one unsafe operation with another.

Reply:
It'll eventually be a `DisjointMut` in the outer loop, though.

Reply:
Ah sure, if it'll make that change cleaner then it makes more sense. Done.